Add support for `license_files` specified in `setup.py`

abravalheri opened the pull request:
As discussed in #465, `bdist_wheel` will behave differently when `license_files` is specified in `setup.py` or in `setup.cfg`.

The main idea of this PR is to make the behaviour consistent across these different possibilities.

I am supposing that the original intention of the `wheel` authors was to also support distributions using `from distutils.core import setup` in addition to `from setuptools import setup`.
In my mind this justifies the choice for `distribution.get_option_dict("metadata")` instead of simply looking up in `distribution.metadata`.
Therefore, I am attempting to preserve this backward compatibility in the PR.

If this is not required the implementation. can be further simplified.

Closes #444.
Closes #465.

abravalheri commented:
The errors in the CI seem to be related to the fact that earlier versions of setuptools (between 45.2.0 and 57.0.0) would process `license_file`/`license_files` differently from what `wheel` expects in its test suite.

When the version of setuptools is updated, the tests seem to run without problems. For example:

```bash
> docker run --rm -it ubuntu:20.04 /bin/bash

apt update -y
apt install -y software-properties-common
add-apt-repository -y ppa:deadsnakes/ppa
apt update -y
apt install -y curl gcc git python3.7 python3.7-dev python3.7-venv
python3.7 -m ensurepip
python3.7 -c 'import setuptools; print(setuptools.__version__)'  # ==> 47.1.0

apt install -y --reinstall python3.7-distutils
curl -sSL https://bootstrap.pypa.io/get-pip.py | python3.7 -

cd /tmp
git clone -q https://github.com/abravalheri/wheel.git -b license_files
cd /tmp/wheel

python3.7 -m pip install --no-binary=:all: .
python3.7 -m pip install .[test] coverage[toml]

python3.7 -c 'import wheel; print(wheel.__version__)'  # ==> 0.37.1
pytest -W always  # ==> Fails: old setuptools behaviour doesn't match wheel's expectations

python3.7 -m pip install 'setuptools==57.0.0'
pytest -W always  # ==> succeds

python3.7 -m pip install -U setuptools
pytest -W always  # ==> succeds
```

agronholm commented:
I think I would prefer the third option. Do you see any downsides for that?

abravalheri commented:
Hi @agronholm, I submitted a few commits simplifying the code with some of the assumptions, just to see how it would work. If necessary I can revert the commits to get the PR to an earlier stage.

The main changes are the following:
- https://github.com/pypa/wheel/pull/466/commits/8055d1cdace5a337fe8bde1209f3c18afd793509 - Here I am always relying on `distribution.metadata` instead of `distribution.get_option_dict("metadata")`.
  - I still keep some of the code for backward compatibility
  - I could see in some versions of setuptools `distribution.metadata.license_files = []` even when the user does not explicitly specify it. This effectively disables the default files and break the test suite expectations.
-  https://github.com/pypa/wheel/pull/466/commits/9fb0b7663b3cb5c512e1922841318fd2b68f2a85 - Here I remove all code for backward compatibility with old versions of setuptools and rely 100% on it already processing `license_files`.
  - I added a few `xfail` to the test suite for old versions of `setuptool` (I also added another `xfail` in https://github.com/pypa/wheel/pull/466/commits/48fe0d6fa2fcb5324957ae7fb7f8865f913cc2a1)

(The other changes mainly target the CI error)

---

> Do you see any downsides for that?

The main downside here is that developers using `python setup.py bdist_wheel` instead of `python -m build` *AND* that run the build in machines with old versions of setuptools installed may have a different set of license files included in the wheel.

This may also happen for people pinning versions of setuptools older than `57.0.0` in their `pyproject.toml [build-system] requires`.

agronholm commented on tests/test_bdist_wheel.py:
```diff
@@ -94,10 +103,29 @@ def test_licenses_deprecated(dummy_dist, monkeypatch, tmpdir):
+        pytest.param(
+            "setup.cfg",
+            "[metadata]\nlicense_files=licenses/*\n  LICENSE",
+            marks=lincense_files_in_old_setuptools("57.0"),
```
```suggestion
            marks=license_files_in_old_setuptools("57.0"),
```

agronholm commented on tests/test_bdist_wheel.py:
```diff
@@ -94,10 +103,29 @@ def test_licenses_deprecated(dummy_dist, monkeypatch, tmpdir):
+        pytest.param(
+            "setup.cfg",
+            "[metadata]\nlicense_files=licenses/*, LICENSE",
+            marks=lincense_files_in_old_setuptools("57.0"),
```
```suggestion
            marks=license_files_in_old_setuptools("57.0"),
```

abravalheri commented:
If you would like, I can also bump `install_requires` to `setuptools >= 57`.

agronholm reviewed (changes requested):
These suggestions fix a typo.

abravalheri commented:
Thank you very much for the review @agronholm.

> If this change indeed drops compatibility for earlier versions, then yes. Also, please add a note to the changelog.

While you can still run `bdist_wheel` with an older version of `setuptools`, the resulting `.whl` may not contain all licenses.

In https://github.com/pypa/wheel/pull/466/commits/81c0bf6d5efee908a406eabaf96213c2d77133c3 I updated the dependencies (and removed the now obsolete `xfail`s).

In https://github.com/pypa/wheel/pull/466/commits/67d20a40899286be6db1bcda4e62f304f93e0dc5, I added some notes to the changelog.

agronholm commented on src/wheel/bdist_wheel.py:
```diff
@@ -432,38 +431,20 @@ def _ensure_relative(self, path):
-        return files
+            files.add(license_file)
+
+        license_files = getattr(metadata, "license_files", None) or []
```
Is the attribute entirely absent if the option hasn't been provided?

abravalheri replied:
I believe the attribute is always present if `setuptools` is used.
Is there any chance `bdist_wheel` runs with `distutils`?

agronholm replied:
> I believe the attribute is always present if `setuptools` is used. Is there any chance `bdist_wheel` runs with `distutils`?

`bdist_wheel` is a setuptools extension so I don't think this could ever happen.

abravalheri replied:
Ok, so I will remove the `getattr` and use the attribute directly. Thanks Alex.

abravalheri replied:
Done in https://github.com/pypa/wheel/pull/466/commits/7545b8a6c849c039587f40b75817292dbcbf3045.

agronholm commented on src/wheel/bdist_wheel.py:
```diff
@@ -432,38 +431,20 @@ def _ensure_relative(self, path):
 
-        if "license_file" in metadata:
+        license_file = getattr(metadata, "license_file", None)
+        if license_file:
```
Is this code path ever triggered on any version of setuptools?

abravalheri replied:
Yes, the user can specify `license_file` instead of `license_files` (it will already result in a warning from setuptools).

I can think about a unit test to add here... (I haven't done that before because the code path was basically inherited from the existing implementation).

agronholm replied:
I'm asking because I looked and did not find this attribute on 45.2.0 or 57.0.0. So it would be nice to check if this gets triggered at all or if setuptools just lumps it into `license_files` on its own.

abravalheri replied:
We can see [here](https://github.com/pypa/setuptools/blob/d680efc8b4cd9aa388d07d3e298b870d26e9e04b/setuptools/dist.py#L642) that the `metadata` object has the attribute in the latest version (I am not sure about 45 or 57).

I will add some unit tests with the intention to activate this path, so we can have more certainty about it.

abravalheri replied:
BTW, we can see in the [same function](https://github.com/pypa/setuptools/blob/2451deee140b043dcb18796eabda370806881035/setuptools/dist.py#L583-L599) that setuptools is already adding `license_file` to `license_files`, so maybe `wheel` don't have to analyse it.

abravalheri replied:
In https://github.com/pypa/wheel/pull/466/commits/16826024e90fe08f3affd02766af807c5f978488, I added a check in the test suite to ensure the code path with the warning is triggered.

In https://github.com/pypa/wheel/pull/466/commits/5b09d963f3c678ffaa0602b7444e338792a4d86e, I removed that code path completely since it seems to already be handled by `setuptools` (the difference is the warning message exact wording and the warning class).

agronholm replied:
> BTW, we can see in the [same function](https://github.com/pypa/setuptools/blob/v57.0.0/setuptools/dist.py#L583-L599) that setuptools is already adding `license_file` to `license_files`, so maybe `wheel` don't have to analyse it again...

I concur. I was just looking at the coverage report and I think the `license_files` property could be eliminated entirely.

abravalheri replied:
Perfect.

I removed this code path.

Please let me know if you prefer to remove entirely the `test_licenses_deprecated` (this way the `wheel` test suite is more independent of setuptools deprecation schedule). Tomorrow morning I can have a look at it.

agronholm replied:
Yeah, we should not be testing setuptools code in the `wheel` test suite. Chuck it out :)

abravalheri replied:
Done in 16c90f4

agronholm commented:
Thanks!
